sam.c

```diff
@@ -3192,8 +3218,10 @@ int sam_format1(const bam_hdr_t *h, const bam1_t *b, kstring_t *str)
 
 // Sadly we need to be able to modify the bam_hdr here so we can
 // reference count the structure.
-int sam_write1(htsFile *fp, const sam_hdr_t *h, const bam1_t *b)
+int sam_write1(htsFile *fp, const sam_hdr_t *hdr, const bam1_t *b)
 {
+    const sam_hdr_t *h = hdr;
+    if (!h) h = fp->bam_header;
```

Comment:
Clearer as
```c
const sam_hdr_t *h = hdr? hdr : fp->bam_header;
if (h == NULL) { errno = EINVAL; return -3; }
```
as in particular passing NULL when no headers are attached should be an error rather than a crash :smile: